src/quorumnet/sn_network.cpp

```diff
@@ -0,0 +1,1019 @@
+
+            while (!idle_workers.empty() && !queue_index.empty()) {
+                size_t i = queue_index.front();
+                queue_index.pop();
+                auto &sock = *(i == 0 ? listener : i == 1 ? self_listener : remotes[i - 2]);
+
+                std::list<zmq::message_t> parts;
+                if (recv_message_parts(sock, std::back_inserter(parts), zmq::recv_flags::dontwait)) {
+                    last_conn_index = i;
+                    queue_index.push(i); // We just read one, but there might be more messages waiting so requeue it at the end
+
+                    // A messge to a worker takes one of the following forms:
+                    //
+                    // ["CONTROL"] -- for an internal proxy instruction such as "QUIT"
+                    // ["PUBKEY", "CMD"] -- for a message send from an authenticated SN (ENCODED_BT_DICT is optional).
+                    // ["", "ROUTE", "CMD"] -- for an incoming message from a non-SN node (i.e. not a SN quorum message)
+                    //
+                    // The latter two may be optionally followed by one frame containing a serialized bt_dict.
+                    //
+                    // The pubkey form supports sending a reply back to the given PUBKEY even if the
+                    // original connection is closed -- a new connection to that SN will be
+                    // established if required.  The routed form only supports replying on the
+                    // existing incoming connection (any attempted reply will be dropped if the
+                    // connection no longer exists).
+
+                    std::string pubkey;
+                    if (i == 1) { // Talking to ourself
+                        pubkey = this->pubkey;
+                    } else {
+                        try {
+                            const char *pubkey_hex = parts.back().gets("User-Id");
+                            auto len = std::strlen(pubkey_hex);
+                            assert(len == 0 || len == 64);
+                            if (len == 64)
+                                pubkey = from_hex(pubkey_hex, pubkey_hex + 64);
+                        } catch (...) {} // User-Id not set, i.e. no pubkey
+                    }
+
+                    if (!pubkey.empty()) {
+                        // SN message which means we want to stick the pubkey on the front.  For a
+                        // connection on the listener we also want to drop the first part (the
+                        // return route) because SN replies have stronger routing.
+                        if (i <= 1) // listener or self: discard the return route
+                            parts.pop_front();
+                        parts.emplace_front(pubkey.data(), pubkey.size());
+
+                        if (parts.size() < 2 || parts.size() > 3) {
+                            SN_LOG(warn, "Invalid incoming message; expected 1-2 parts, not " << (parts.size() - 1));
```

Comment:
This is a bit misleading because the pubkey was just shifted onto the message parts but wasn't part of the incoming message.  I can shift that down to after the check to make it seem less weird.